core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -455,13 +521,35 @@ public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec update
     return new TableMetadata(ops, null, uuid, location,
         System.currentTimeMillis(), nextLastColumnId.get(), freshSchema,
         specId, builder.build(), ImmutableMap.copyOf(newProperties),
-        -1, snapshots, ImmutableList.of());
+        -1, snapshots, ImmutableList.of(), previousMetadata, removedPreviousMetadata);
   }
 
   public TableMetadata updateLocation(String newLocation) {
     return new TableMetadata(ops, null, uuid, newLocation,
         System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
-        currentSnapshotId, snapshots, snapshotLog);
+        currentSnapshotId, snapshots, snapshotLog, previousMetadata, removedPreviousMetadata);
+  }
+
+  public TableMetadata addPreviousMetadata(String previousFile, long timestampMillis) {
```

Comment:
What is the purpose of the timestamp here?

Reply:
Just to make sure we don't delete latest metadata. Also it might be handy, if we decided to do time based expiry in future.